<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Daniel Greene" />

<meta name="date" content="2021-02-15" />

<title>IntervalSurgeon</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">IntervalSurgeon</h1>
<h4 class="author">Daniel Greene</h4>
<h4 class="date">2021-02-15</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p><code>IntervalSurgeon</code> presents functions for intersecting, overlapping, piling and annotating integer-bounded intervals. Underlying algorithms are written in C++ for efficiency where appropriate (with the help of the <code>Rcpp</code> package). A typical use case would be for manipulating genomic intervals.</p>
<p>For the purposes of this package, intervals are represented as two-column integer matrices where the inclusive start points are in the first column and the exclusive end points are in the second column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(IntervalSurgeon)

starts &lt;-<span class="st"> </span><span class="dv">3</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>
ends &lt;-<span class="st"> </span><span class="dv">5</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>

intervals &lt;-<span class="st"> </span><span class="kw">cbind</span>(starts, ends)</code></pre></div>
<p>The lengths of the intervals are therefore: <code>intervals[,2]-intervals[,1]</code>.</p>
<p>A key concept in <code>IntervalSurgeon</code> is breaking ranges which contain intervals into 'sections' delimited by the unique start/end points in the set. The sections for a set of intervals <code>x</code> is therefore a two-column <code>matrix</code> representing a set of intervals which partitions the range of <code>x</code> by the sorted start and end points. The sorted start and end points can be obtained using the <code>breaks</code> function (so named to reflect start/end points of intervals frequently being referred to as 'breakpoints' in genomics), which is equivalent to: <code>sort(unique(as.integer(intervals)))</code>). The sections can be computed from the sorted start and end points using the <code>sections</code> function.</p>
<p>One can use the <code>depth</code> and <code>pile</code> functions respectively to obtain the depth of intervals over their 'sections' (within which the depth is constant), and the row indices of original intervals which cover each section.</p>
<p>The <code>flatten</code> function returns a non-touching, non-overlapping set of intervals (as a <code>matrix</code>) which overlap at least one of a given set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sectioned &lt;-<span class="st"> </span><span class="kw">sections</span>(<span class="kw">breaks</span>(intervals))
flattened &lt;-<span class="st"> </span><span class="kw">flatten</span>(intervals)
depths &lt;-<span class="st"> </span><span class="kw">depth</span>(intervals)
piles &lt;-<span class="st"> </span><span class="kw">pile</span>(intervals)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAEgCAMAAABcujGyAAAAZlBMVEUAAAAAADoAAGYAAP8AOpAAZrY6AAA6ADo6kJA6kNtmAABmADpmZmZmkJBmtrZmtv+QOgCQkGaQtpCQ2/+2ZgC225C2/7a2///T09PbkDrb/7bb////AAD/tmb/25D//7b//9v////UndH6AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAJ+0lEQVR4nO3dD1OjRgCGcbSaq7n28Jq0NmdJ5Pt/ye4uAYNK2F1Y9hWf34xzScANf54BEqelqAFhRe4FAK4hUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgjUEgj0GCnbfn6pCqHZ3w/N4IR6CTj+RHoNAQ6CYGmRqDBbHKn7Y9tUZTHTVHcPdeHoigezITvP80D22N1s6v3zWM7t52toNMoBBqsCdQkeLjZueOj+dc0+GBevHt+ebw3s+zvnvf3zQQzx3FjZqooNAqBBmsCNUdME173uK5un9yDw+2TfeX0fedmsHObSbmX+fMi0GBNlM3Z2/5rz+ddrfXrK+agWZTNTBQajUCDvQu0aDSB1ubcvm+uS2//bo6g9csj16CxCDTYx0fQuv3AXt3+4z4Xld0p3k3ct3MhCIEGexto90VS8+C0/cOc0d11Z1W8TuXrpjgEGqwf6EPzYd0eIc8N7ov79uBZPLgPSXZ6xRE0CoEGuwzU1Hj+HtR9eHeBNl8omddududq7VUqfcYhUEgjUEgjUEgjUEgjUEgjUEgbDrQYtuDy4YsbjO1KnwSKxRAbpBEopBEopE0P9NrFKlexmGhyKlF9Eig8kQqkESikESikLRdo3MUqV7Ff3GI7euY+CfSLYEdDGoFCGoFCml6gc1+schX7qcntpsX6lFtzfITdBGkECmkECmmfNtDlrlWvyL0RvoDPuo1zp9nIvRW+ALYxpBEopBEopK0+0NyXqSNybx55a99CuQMck3v7yGMLQRqBQhqBQhqBnuW+GI2We8Mltvb185U7s3i5t1xia18/fHIECmkECmkEGib3FWcKubfpVdpLJyd3S0nk3qhXaS8dvjwChTQChTQCTSP3deWyUm7IhGN/YbmLWVjKLZlwbGAyAoU0AoU0As0r99WjjMENtOTewFu5s9AxuIWW3B1AKAKFNAKFNAKFNAKFNAKFNAKFNAKFNAKFNAKFNAKFNAKFNAKFNAKdR1XWp205PtvNboGFWRMCnYVPnBaBhiLQWRBoKgQa5LgpisK2eDD/PtTnB/f25btfLtKqmXDa/ti+n/PmJ4EGItAQx01pEyzrgwntuDHd2Qen7YM9grqjqJ142t6bHzPhcP5xc+5vn8xUAg1EoCEqE5llkmyeNQ+aU7z9eXl0E252boLpuZvTtV3vCTQQgYY4bZtCm0tJ01yT3WugzXMXZule7eZs2uYaNBSBBnl5dNeg1fm/9CqP35rg+oGaR12g7ZwHAo1CoMHMabrrzPMIWrdXBwQaikCDtcfH5vHgNWjZn7NJ90CggQg0hDsAVufP5u4Tj33w8nhvQ33zKb4JtD8nn+KDEWgQe0npGrPfbrqTtvse1CT47nvQc6C9OfkeNBiBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQtpQoL/9NvgrvUlX5ov6Jc/xot72ypSot/Ud3He8GOsY4coQw4ECCyJQSAsPdOg3OMWHDu47Xox1jBB1ivcbjEDD3jZuvrWPQKBjUwg06wgRgQISCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCBTSCDTYoSh+b2+ybVVl8zOCW8VHIdBQ3Z23u+dl7/kQAo1CoKHeBkmgSRFooOOmKG7/sUHui6Io3XPr7tndGd7dKv7H1k6p2xe4VfwEBBqqPWLu7014JrruCGqfHDf2/G8eHM4/9oV6f/tUVwWBxiDQUOcgT9939nBadoHaa1NzIr99cg/MlO4FO5eplEBjEGioi2vOyp7J2+fNNaYLs3RzdS+YRmuuQSMRaKg2SHNdefv3xRG0aq5Fi9dA2xcOBBqPQEOdg3Tn7ctTfBdg/whau7N8TaCRCDTUOUhXXVVcXoOWrzO4q9T2heYa9ECgMQg01MUR9LQtHs5f3D+cCzQfhdpAuxfcAz7FxyHQUBfXoDc7m9++uHvet9+DmuNqF2j7At+DTkCgkEagkEagkEagkEagkEagkEagkEagkEagkEagkEagkEagkEagkEagkEagkEagkDYU6H//Df5Kb9KV+QBfwxkNB3qF73yALwKFtPBAh36DUzxmR6CQFh4oIIFAIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1Ap/j4/ztf+d16Dj4IdIoPA/W9NyJ8EOgUBJocgcbq/r/z5ztyHr/9tbEP7M07735t/9y4+3W6W3kSazwCjdTdf7O9I6dJsby4L00zxd2BpqLQeAQap7v/5sUdOe2Dw91zd+PO7i6IiEegcbq7x3V35GySrZr7JL3ezItCpyHQON39N7s7ch6/7ep3gdYvj1yDTkKgcd4cQWv7IemjQC1uxD0Bgcbp7r95cUdOew26b69BLwLlO6cJCDRSd//N9o6cx417pTzfvPPijsfc53gCAo3V+x7UnPCPm/NXn/bmnb/aI6i9RKXPCQh0Ls1JHzMj0LkQaBIEOhcCTYJAIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AIY1AgxR9vvOltODa57D29ZtZ8e+lK4H+68d3visjLLj2Oax9/WZGoEtb+/rNjECXtvb1mxmBLm3t6zczAl3a2tdvZgS6tLWv38wIdGlrX7+ZEejS1r5+MyPQpa19/WZGoEtb/fp5/o3Q9w+L/Tg850sa6HSZdo0f7aWbbriAwnO+K0FdGdxvuDkCnT5Cpl3jR3vppiPQ8REy7Ro/2ks3HYGOj5Bp1/jRXrrpCHR8hEy7xo/20k1HoOMjZNo1frSXbjoCHR8h067xo7100xHo+AiZdo0f7aWbjkDHR8i0a/xoL910BDo+QqZd40d76aYj0PERMu0aP9pLNx2Bjo+Qadf4WWzpfP/8O8PflnuG94vnfFdyWEmg06XsJuHY/TfqbRPP+abvF9+GvnKg00dI2U3Csftv5LlGBBpGYoSU3SQcu/9GnmtEoGEkRkjZTcKx+2/kuUYEGkZihJTdJBy7/0aea0SgYSRGSNlNwrH7b+S5RgQaRmKElN0kHLv/Rp5rRKBhJEZI2U3Csftv5LlGBBpGYoSU3SQcu/9GnmtEoGEkRkjZTcKx+2/kuUYEGkZihJTdDE7w/MuW95/DemvkOd/0rUqgS4wwg/BAryyP53zDm+Ha4BHDRc1HoEIjDB+ECXTC2xLoXCMQ6NgUAs06AoGOTSHQrCMQ6NgUAs06AoGOTSHQrCMQ6NgUAs06AoGOTSHQrCMQ6NgUAs06AoGOTSHQrCPEBAosKDhQQAGBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQhqBQtr/OsrecetD3VsAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
</div>
<div id="detached-intervals" class="section level2">
<h2>Detached intervals</h2>
<p><code>IntervalSurgeon</code> provides functions which are optimised for dealing with detached (i.e. non-overlapping and non-touching) intervals which are sorted and non-empty. Here, we generate two such sets of intervals:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x_starts &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>
x &lt;-<span class="st"> </span><span class="kw">cbind</span>(x_starts, x_starts <span class="op">+</span><span class="st"> </span><span class="dv">5</span>)

y_starts &lt;-<span class="st"> </span><span class="dv">20</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">5</span> <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
y &lt;-<span class="st"> </span><span class="kw">cbind</span>(y_starts, y_starts <span class="op">+</span><span class="st"> </span><span class="dv">7</span>)</code></pre></div>
<p>We can determine that they are indeed detached, sorted and non-empty using the <code>detached_sorted_nonempty</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">detached_sorted_nonempty</span>(x)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>The <code>IntervalSurgeon</code> functions for operating on such sets of detached, sorted, non-empty intervals are analogues of the set operation functions in the <code>base</code> package, namely: <code>intersect</code>, <code>union</code> and <code>setdiff</code>. Here, the function names are in plural (i.e. with extra s's on the end).</p>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAGACAMAAABFpiBcAAAAhFBMVEUAAAAAADoAAGYAOpAAZrYAZv8A/2Yra6M6AAA6ADo6Ojo6OpA6kNtMAABMh75mAABmADpmZmZmtrZmtv9ro76QOgCQkGaQ27aQ2/+jayu2ZgC2/7a2//++h0y+o2u+voe+vr7MAP/M/wDbkDrb/9vb////AAD/tmb/25D//7b//9v///8C2BbuAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAMhElEQVR4nO3YjVYiVwJF4fIniZmhm54mZhzHdJswBrHe//2mqkBkIyhg0Zyr+6yVRLmYT1h7FWBVb9jtpoOdzjffa83J7cvH8h+Ex6q3/RKvnZf6HMkfkccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDLQ/fnx6U0+Hg2Pxz85L5TED7Y9/uBzU47O7Y/HPzkvlMQPtkZ+c/vfL1fH41fNSecxA++Svq/Nj8ivnpfKYgfbJT6rRMfmV81J5zEB75B8u/9V8Tjoav3peKo8ZaI/8+Ox/l4Pj8avnpfKYgfbHT4ej5nPSzbH4Z+el8piB9sdfn901r/Lnx+KfnZfKYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GY5sC/fvvl3+HLc83uut+/vblY/mPwXMGKh/Gc5sDfW2v/ZK7//zt235c/n3wnIHKh/GcL/HyYTxnoPJhPGeg8mE8Z6DyYTznH+rl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zED746/Pm3+Nz4/FPzsvlccMtD9+cnpTP1yOjsU/Oy+Vxwy0P346HNX3v14di392XiqPGWiPfPMaPz67Oxq/el4qjxloj3zzGn89OB6/el4qjxloj/z0y7+/XB2PXz0vlccMtEf+4fIfs1f4D/noe+MxA+2TH1eDY/Ir56XymIH2yc8/w3/QR98Xjxlon/z9P++Oya+cl8pjBtonPx4clV85L5XHDLQ//v5i/hHpQz76/njMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jMQOXTeMxA5dN4zEDl03jssIH+9dcOP487y39cHjNQ+TQeO3SgO01efnUGKp/GY77Ey6fxmIHKp/GYgcqn8ZiByqfx2GEDfeFea05uXz6W/yA8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGagu/DT4ejpfDLamn921yIf/Y/iMQPdl1/E+jr//K7FP/pD8piB7ssb6KF4zEB34dvSpsOvw6oa3V9U1dldPa6qatAcfPm9Or1pb6vaFue33nZfnM/uujjcn9/xvNQnHzPQXfhZoCdX9X9OrrrL4rj5+v5i0NzYxHp/0dwyaSIcn/zW3XrbHk+Hg/aui8M38Duel/rkYwa6Cz8LtLk2frsYLb6uJ6c33RfNf+f3Gtx2333vjmc/9nj4Fn7H81KffMxAd+FnUTZXwe/DLtBJc4Ws24tjd+N0OIuwufW2u/XbxWjpx5YLLfLR/ygeM9Bd+GeBVrPNAq0fLru3mYtbv/16tfixx8O38Duel/rkYwa6C7/+Clovf06/PrmaXUHr7o0AT68f71/ko/9RPGagu/CrgS7Kewp02p3Mfn75PSi/KPLR/ygeM9BdeAY6mH2Kb6+L3Y3d9bT9V/spvr21+xT/cHne3nVx+AZ+x/NSn3zMQHfhlwOtrx//Dtp8+pldGts3n12C81vnfwetu7suDvfndzwv9cnHDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDFQ+jccMVD6NxwxUPo3HDhvozz9vOl530vtzJF8kjxmofBqPHTrQzXv5fyn/cXnMQOXTeMyXePk0HjNQ+TQeM1D5NB4zUPk0HjtsoC/ca81J78+RfJE8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGagPfHT4eiY/LrzUnnMQOXTeMxA5dN4zED35LuX9OZf0+HXYVWN5i/xk6qqBs3tn7obb+v7i6o97J/f5rzUJx8z0D35p0BPrurxyVX3/aQr9bwJtLvxt/uL+Y2989ucl/rkYwa6J/8UaHPFbEJsv3+4bL6uJ02tn7obP09Obw7Eb3Ne6pOPGeie/FOgT//tLphdrZ+6Gz9PhxsKLf3RH5THDHRPfnOg7TfzQJtrqu9Bd+YxA92T3+oK2t33unk/2ju/zXmpTz5moHvyy+89HwNdeg+6FCj/hN8Tv815qU8+ZqB78g+XZ3fNCzgCXfoUPwt00l48J15Bd+MxA92Xnw6r6itf4pf+Djq/grbfr+mz/Ed/SB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB7bFOhPP738O2x5vssvufy/lP+4PGeg8mE8Z6DyYTy3OdDX9tovucfk5VdnoPJhPOdLvHwYzxmofBjPGah8GM/5h3r5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4zUPk0HjNQ+TQeM1D5NB4z0J74yejxi5Or6fBzXY+r6pfmn8GP4dedl/rkYwbaDz8dLgXank+Hg/r7cH2e7+7R98tjBtoPvybQURPoaPV+B+LXnpf65GMGuh9/f1FVVZtf8yrevIy3357dtd+c/N69xHfnzU5vFvfsk9/qvNQnHzPQvfhvF01yk6a7cXO9vL8YzK6g16c3zY2z96CPV9D7x3v2yb/vJx8z0L34P5oU2027d5mT05vpY4v1NQOdzO/ZK/++n3zMQPfivw9n3XVvONsyp4sWJwx0OlxXaNmP3kDz+YfL7p3lZPZOs+oCHa8LtJ7fs19+q/NSn3zMQPfnmxfz2RW0nn2KX3sFref37J9/9bzUJx8z0P35psHFX5ee3oOO1wQ6Xf170zt49IfjMQPdi/+jvSS2l8/2U/zsc9Gga/PZp/jJ4z375N/3k48Z6H58++azq27c/bGziZR/B326gi7u2Se/1XmpTz5moPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8ZqDyaTxmoPJpPGag8mk8tinQP/98+XfY7vyFe738S8p/XJ4zUPkwnjNQ+TCe2xzoa3vtl3zlXq89R/IflecMVD6M53yJlw/jOQOVD+M5A5UP4zn/UC+fxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmMGKp/GYwYqn8ZjBiqfxmP/B4lmBAZenAsSAAAAAElFTkSuQmCC" style="display: block; margin: auto;" /></p>
<p>For a given set of sorted, non-overlapping intervals, some of the start points might be the same as the end points of adjacent intervals. These intervals are therefore 'touching' and can be 'stitched' together using the <code>stitch</code> function. If there were overlaps, the <code>flatten</code> function can be used to generate a set of sorted disjoint intervals. Note that <code>flatten</code> will also stitch touching intervals, although the <code>stitch</code> function is faster (albeit requiring intervals to be sorted).</p>
</div>
<div id="finding-overlaps-between-sets-of-intervals" class="section level2">
<h2>Finding overlaps between sets of intervals</h2>
<p>Information about overlaps between sets of intervals can be obtained by 'joining' the sets. This is analogous to an SQL inner-join of two tables, and can be performed on sets of intervals using the <code>join</code> function. This function returns a matrix containing all overlaps of intervals from each set. Each row in the returned matrix corresponds to a specific overlap of intervals with one from each of the sets passed to the function. The <code>n</code>th element in a row contains the row index of the interval in the <code>n</code>th set of intervals passed to the function. Depending on the value of the <code>output</code> argument, there may two additional columns giving the start and end coordinates of the overlap (the default: <code>output=&quot;intervals&quot;</code>, no extra columns (<code>output=&quot;indices&quot;</code>) or one additional column giving the row index of the 'section' of the complete set of intervals (<code>output=&quot;sections&quot;</code>, see <code>?sections</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dv">3</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">8</span>, <span class="dv">5</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">8</span>)
y &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dv">4</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, <span class="dv">7</span><span class="op">*</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)
join_xy &lt;-<span class="st"> </span><span class="kw">join</span>(x, y)

<span class="kw">head</span>(join_xy)</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,]    1    1    4    5
## [2,]    2    1    6    7
## [3,]    2    2    8   10
## [4,]    3    2    9   14
## [5,]    3    3   12   15
## [6,]    4    2   12   14</code></pre>
<p>One common task would be to tag intervals with overlapping intervals, perhaps from a different set. For example, this might be useful for tagging a set of genomic intervals with the names of genes which they overlap. The <code>annotate</code> function is provided for this exact purpose.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dv">0</span>, <span class="kw">c</span>(<span class="dt">a=</span><span class="dv">10</span>, <span class="dt">b=</span><span class="dv">20</span>, <span class="dt">c=</span><span class="dv">30</span>))
y &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">c</span>(<span class="dt">A=</span><span class="dv">0</span>, <span class="dt">B=</span><span class="dv">10</span>, <span class="dt">C=</span><span class="dv">20</span>), <span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">15</span>, <span class="dv">25</span>))</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAFQCAMAAABXmDxzAAAAaVBMVEUAAAAAADoAAGYAOpAAZrYA//86AAA6ADo6kNtMAABmAABmADpmZjpmkJBmtv9ro76Ho2uQOgCQ2/+jayu2ZgC225C2//++h0y+o2u+vr7bkDrb////AAD/tmb/trb/25D//7b//9v///9A+BN9AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAJQElEQVR4nO3dXVPcyBlA4bEXY4ckbHCId80u4eP//8gwgHNBDU5popF0+pz3wje71EO3To1Gsqt69/jOfH/vP8w8gzmDLWd9Zzf5J2aewZzBlrO+U6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TjrBnq92+0ul4AKlOqsGuj12ePjPz98XUAqUKqzZqD3f31q8/fzZT5CC5TprP0d9Hape3yBMp1VA73Z7T7+o0/QnI0Gerdvs1t8zs+cNQO9/fjt8fG3bvE5Gw10/wl6/5fdl9NLjwVKddb+Dvrh1+tl3jMVKNNZ+yl+NGew5azvFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE47wX6558n+k3ezkI7sNR6Vr+gozkFOu+sfkFHc94PdLQ50c6+mdUv6GhOgc47q1/Q0Zxu8fPO6hd0NKdA553VL+hoTq+ZkIzHKVAk43EKFMl4nAJFMh6nQJGMxylQJONxChTJeJwCRTIep0CRjMcpUCTjcQoUyXicAkUyHqdAkYzHKVAk43EKFMl4nAJFMh6nQJGMxylQJONxChTJeJwCRTIep0CRjMcpUCTjcQoUyXicAkUyHqdAkYzHKVAk43EKFMl4nAJFMh7HEOjzmaAFynQKdN4ZaNu24RTovDPQtm3DcQT69/PdWGfWehxHoB++Pv5+vkihA23bNhxHoE9tfr/Zn05/8hlo27bhOALdfwe9XeTY74G2bRuOItDPT2l+f/7z5DPQtm3DUQTaJyjXcQT6/B30lz9OT420bdtwHIE+fXj+trs8vTTUtm3DcQS6fw+6SJ8jbds2HEOgSzqDLWd9p0CRjMcpUCTjcQoUyXicAkUyHqdAkYzHsRzHPVignstToEjGc3neD7TZ9JwolLdToM1xc6JQ3s6GAz3Rb/J2Vt8BJOO5PAWKZDyXp9dMSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6Bzjk3u93ubAlorG37mVOgM87Nh6/f7y++LCANtW0/dQp0vtm3OdBytuEU6Hxzd3450nK24RTofHP3+etIy9mGU6DzTZ+gJ3AKdL7pO+gJnAKdcfZP8Q9Xi7xnGmnbfuoU6JzTe9DZnQJFMh6nQJGMxylQJONxChTJeJwCRTIep0CRjMfpOG4kM5rzfm4FimRGc44JtGkWnAJtNj3TA33vJ2aewe5Vgy1ny7f4E/0qb2ewnR5sOdsNdLQdKFCoU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDzOjIFe7445KX31HUAyHme+QO8+/+3jt6k/tIEdQDIeZ75Abz7+6/xy6g9tYAeQjMeZLdCHq7OjzlFdfQeQjMeZLdDbD1/3p/1O/KkN7ACS8TizBXr9yx/PB6ZPndV3AMl4nLkCvTvft3k9/TFp9R1AMh5nrkD3J6XvZ/Jj0uo7gGQ8zkyBvj4f3V9MfkxafQeQjMeZKdDb18ej6Y9Jq+8AkvE4MwW6f0Taz8tX0Smz+g4gGY8zT6B3/31FP/kxafUdQDIep38sgmQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmP816gnz6d6DdZySlQqPN+oEvNiVb8dgqU6RQokvE43eKRjMcpUCTjcQoUyXicXjMhGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BHp7r56NxJx/suNXlcJ0CPTwvZzvqjx5d3ynQw/MS6P2F/OjR9Z0CPTw/Ar38n//n/8UcPRqnQA/P6y3+7MTM0aNxCvTwvDwk7V5PGT8Zc/RonAI9PC+foLe7qR+hG10O1ynQw/MS6OP1x28nZY4ejVOgh+c10JsCXdkp0MPz4xN06pfQjS6H6xTo4fnxHbTXTCs7BXp4+qvOjTgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPYwl0oWO/RztdfH3HE+hCs8xyRrs8BVqgUMcT6EjMcJenQAsU6hQokRnu8hRor5mgToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkpjsPV8ecLr7+egoUyUx2bndfnv68nnx03urrKVAkM9W5O38+dPThCnc4boEimanOzWuY/554uvj66ylQJDPRebg6W8Q5fgpUHej9xZdFnOOnQAt0Aef4KVB1oN3ij5/BnI0u58dD0u3u8qTO0VOg7kB7zXT0DOZsdTnPL+ofrnpRP3kGcza7nPuL/qrzqBnMGWw56zsFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIpmljv1e/3jxAkUynxabZdZToAVaoMfNYE63+OOmQAt0y06BFuimnQIdLFCPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4BYpkPE6BIhmPU6BIxuMUKJLxOAWKZDxOgSIZj1OgSMbjFCiS8TgFimQ8ToEiGY9ToEjG4xQokvE4/wHU9zWsSe0NgAAAAABJRU5ErkJggg==" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">annotate</span>(x, y)</code></pre></div>
<pre><code>## $a
## [1] &quot;A&quot;
## 
## $b
## [1] &quot;A&quot; &quot;B&quot;
## 
## $c
## [1] &quot;A&quot; &quot;B&quot; &quot;C&quot;</code></pre>
</div>
<div id="use-with-genomic-intervals" class="section level2">
<h2>Use with genomic intervals</h2>
<p>Genomic intervals are often represented in R as <code>data.frame</code>s with columns corresponding to chromosome name, start position and end position. Obviously intervals do not intersect if they are on different chromosomes, so in order to manipulate such intervals with <code>IntervalSurgeon</code>, genome-wide operations must be performed chromosome-at-a-time. Using <code>split</code> to create a list of chromosome specific <code>data.frames</code>, or looping over the names of chromosomes and subsetting the original table, before <code>cbind</code>ing/<code>as.matrix</code>ing the start and end columns then makes the data accessible to the <code>IntervalSurgeon</code> functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">regions_annotated_with_genes &lt;-<span class="st"> </span><span class="kw">lapply</span>(
    <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">22</span>, <span class="st">&quot;X&quot;</span>, <span class="st">&quot;Y&quot;</span>),
    <span class="cf">function</span>(chromosome) {
        regions_on_chr &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(regions[regions<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span>chromosome,<span class="kw">c</span>(<span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>)])
        genes_on_chr &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(genes[genes<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span>chromosome,<span class="kw">c</span>(<span class="st">&quot;start&quot;</span>,<span class="st">&quot;end&quot;</span>)])
        <span class="kw">annotate</span>(regions_on_chr, genes_on_chr)
    }
)</code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
